coaster/roles.py

```diff
@@ -0,0 +1,353 @@
+
+        id = db.Column(db.Integer, primary_key=True)
+        name = db.Column(db.Unicode(250))
+        set_roles(name, rw={'owner'})  # Specify read+write access
+
+        title = db.Column(db.Unicode(250))
+        set_roles(title, write={'owner', 'editor'})  # Grant 'owner' and 'editor' write but not read access
+
+        @set_roles(call={'all'})  # 'call' is an alias for 'read', to be used for clarity
+        def hello(self):
+            return "Hello!"
+
+        # Your model is responsible for granting roles given a user or
+        # user token. The format of tokens is not specified by RoleMixin.
+
+        def roles_for(self, user=None, token=None):
+            # Calling super give us a result set with the standard roles
+            result = super(RoleModel, self).roles_for(user, token)
+            if token == 'owner-secret':
+                result.add('owner')  # Grant owner role
+            return result
+"""
+
+from __future__ import absolute_import
+from functools import wraps
+import collections
+from copy import deepcopy
+from sqlalchemy import event
+
+__all__ = ['RoleAccessProxy', 'RoleMixin', 'set_roles', 'declared_attr_roles']
+
+
+class RoleAccessProxy(collections.Mapping):
+    """
+    A proxy interface that wraps an object and provides passthrough read and
+    write access to attributes that the specified roles have access to.
+    Consults the ``__roles__`` dictionary on the object for determining which roles can
+    access which attributes. Provides both attribute and dictionary interfaces.
+
+    Note that if the underlying attribute is a callable, calls are controlled
+    by the read action. Care should be taken when the callable mutates the
+    object.
+
+    :class:`RoleAccessProxy` is typically accessed directly from the target
+    object via :meth:`~RoleMixin.access_for` (from :class:`RoleMixin`).
+
+    Example::
+
+        proxy = RoleAccessProxy(obj, roles={'writer'})
+        proxy.attr1
+        proxy.attr1 = 'new value'
+        proxy['attr2'] = 'new value'
+        dict(proxy)
+
+    :param obj: The object that should be wrapped with the proxy
+    :param roles: A set of roles to determine what attributes are accessible
+
+    """
+    def __init__(self, obj, roles):
+        self.__dict__['_obj'] = obj
+        self.__dict__['_roles'] = roles
+
+        # Read and write access attributes for the given roles
+        read = set()
+        write = set()
+
+        for role in roles:
+            read.update(obj.__roles__.get(role, {}).get('read', set()))
+            write.update(obj.__roles__.get(role, {}).get('write', set()))
+
+        self.__dict__['_read'] = read
+        self.__dict__['_write'] = write
+
+    def __repr__(self):  # pragma: no-cover
+        return 'RoleAccessProxy(obj={obj}, roles={roles})'.format(
+            obj=repr(self._obj), roles=repr(self._roles))
+
+    def __getattr__(self, attr):
+        if attr in self._read:
```

Comment:
Should this be:

```python
    if attr in self._read or attr in self._write:
```

so, that a user with write-access automatically gets read-access.